onnxruntime/core/providers/cuda/tensor/transpose_impl.cu

```diff
@@ -7,6 +7,167 @@
+      break;
+    case sizeof(int32_t):
+      _Transpose3DKernel<int32_t><<<grid_size, block_size, 0>>>(
+          input_shape, input_strides,
+          reinterpret_cast<const ToCudaType<int32_t>::MappedType*>(input_data),
+          reinterpret_cast<ToCudaType<int32_t>::MappedType*>(output_data));
+      break;
+    case sizeof(int64_t):
+      _Transpose3DKernel<int64_t><<<grid_size, block_size, 0>>>(
+          input_shape, input_strides,
+          reinterpret_cast<const ToCudaType<int64_t>::MappedType*>(input_data),
+          reinterpret_cast<ToCudaType<int64_t>::MappedType*>(output_data));
+      break;
+    default:
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Type not supported for transpose on CUDA. Element size was ",
+                              element_size);
+  }
+
+  return Status::OK();
+}
+
+template <typename T, int element_size>
+__global__ void _Transpose4DKernel(const TArray<int64_t> input_strides, const T* input_data,
+                                   const TArray<int64_t> output_strides, T* output_data,
+                                   CUDA_LONG N) {
+  // output coordinates will be: blockIdx.y, blockIdx.x, threadIdx.y, threadIdx.x
+  CUDA_LONG input_index = (blockIdx.y * input_strides[0] +
+                           blockIdx.x * input_strides[1] +
+                           threadIdx.y * input_strides[2]) / (4 * sizeof(int) / element_size) +
+                           threadIdx.x * input_strides[3];
+
+  CUDA_LONG output_index = (blockIdx.y * output_strides[0] +
+                            blockIdx.x * output_strides[1] +
+                            threadIdx.y * output_strides[2]) / (4 * sizeof(int) / element_size) +
+                            threadIdx.x * output_strides[3];
+
+  const int4* v_input = reinterpret_cast<const int4*>(input_data);
+  int4* v_output = reinterpret_cast<int4*>(output_data);
+
+  if (input_index < N && output_index < N) {
+    v_output[output_index] = v_input[input_index];
+  }
+}
+
+bool canDoTranspose4D(const cudaDeviceProp& prop,
+                      size_t element_size,
+                      int32_t rank,
+                      const std::vector<int64_t>& input_dims,
+                      const std::vector<size_t>& permutations) {
+  if (rank == 4 &&
+      // the permutations is not on the last dimension.
+      permutations[rank-1] == (rank - 1)) {
+
+    // The block size will be set based on the last two dimensions of 4D tensor.
+    // the number threads per block will be calculated as below.
+    int num_elements_per_thread = 4 * sizeof(int) / element_size; // int4 is used in the kernel to access data.
+    int64_t num_elements_in_last_two_dimensions = input_dims[rank-2] * input_dims[rank-1];
+    int64_t num_threads_per_block = num_elements_in_last_two_dimensions / num_elements_per_thread;
+
+    if (((num_elements_in_last_two_dimensions & (num_elements_per_thread - 1)) == 0) &&
+        num_threads_per_block <= prop.maxThreadsPerBlock &&
```

Comment:
>maxThreadsPerBlock [](start = 38, length = 18)

what's the difference between prop.maxThreadsPerBlock and GridDim::maxThreadsPerBlock, and prop.warpSize and GPU_WARP_SIZE and 32 (as hard-coded in the comment below)?

Reply:
On Nvidia GPU, prop.maxThreadsPerBlock is 1024, GridDim::maxThreadsPerBlock is 256 which we chose for our kernels. prop.warpSize is 32 but it is not a constant during compiling time. GPU_WARP_SIZE is constant, which can benefit optimizations such as loop unroll.

Reply:
is there a reason to prefer prop.warpSize (e.g. here) over GPU_WARP_SIZE?

Reply:
GPU_WARP_SIZE is a constant at compiling time and power of 2, then the compiler can optimize it with shift instruction for multiplicaiton.